62.CAD/Hatch.cpp

```diff
@@ -0,0 +1,782 @@
+	double t3 = (4 * p0x * (p1y * p1y)) - (4 * p0x * p1y * p2y) - (4 * p1x * p0y * p1y) + (8 * p1x * p0y * p2y) - (4 * p1x * p1y * p2y) - (4 * p2x * p0y * p1y) + (4 * p2x * (p1y * p1y)) - (2 * p0x * p0y * p2y) + (2 * p0x * (p2y * p2y)) + (2 * p2x * (p0y * p0y)) - (2 * p2x * p0y * p2y);
+	double t4 = -(4 * p0x * p1x * p1y) - (4 * p0x * p1x * p2y) + (8 * p0x * p2x * p1y) + (4 * (p1x * p1x) * p0y) + (4 * (p1x * p1x) * p2y) - (4 * p1x * p2x * p0y) - (4 * p1x * p2x * p1y) + (2 * (p0x * p0x) * p2y) - (2 * p0x * p2x * p0y) - (2 * p0x * p2x * p2y) + (2 * (p2x * p2x) * p0y);
+	double t5 = (4 * p0x * p1x * p1y * p2y) - (4 * (p1x * p1x) * p0y * p2y) + (4 * p1x * p2x * p0y * p1y) - ((p0x * p0x) * (p2y * p2y)) + (2 * p0x * p2x * p0y * p2y) - ((p2x * p2x) * (p0y * p0y)) - (4 * p0x * p2x * (p1y * p1y));
+
+	// "Slam" the other curve onto it
+
+	float64_t2 A = second.p[0] - 2.0 * second.p[1] + second.p[2];
+	float64_t2 B = 2.0 * (second.p[1] - second.p[0]);
+	float64_t2 C = second.p[0];
+
+	// Getting the quartic params
+	double a = ((A.x * A.x) * t0) + (A.x * A.y * t1) + (A.y * t2);
+	double b = (2 * A.x * B.x * t0) + (A.x * B.y * t1) + (B.x * A.y * t1) + (2 * A.y * B.y * t2);
+	double c = (2 * A.x * C.x * t0) + (A.x * C.y * t1) + (A.x * t3) + ((B.x * B.x) * t0) + (B.x * B.y * t1) + (C.x * A.y * t1) + (2 * A.y * C.y * t2) + (A.y * t4) + ((B.y * B.y) * t2);
+	double d = (2 * B.x * C.x * t0) + (B.x * C.y * t1) + (B.x * t3) + (C.x * B.y * t1) + (2 * B.y * C.y * t2) + (B.y * t4);
+	double e = ((C.x * C.x) * t0) + (C.x * C.y * t1) + (C.x * t3) + ((C.y * C.y) * t2) + (C.y * t4) + (t5);
+
+	return hatchutils::solveQuarticRoots(a, b, c, d, e, 0.0, 1.0);
+}
+
+double Hatch::QuadraticBezier::intersectOrtho(double coordinate, int major) const
+{
+	// https://pomax.github.io/bezierinfo/#intersections
+	double points[3];
+	for (uint32_t i = 0; i < 3; i++)
+		points[i] = major ? p[i].y : p[i].x;
+
+	for (uint32_t i = 0; i < 3; i++)
+		points[i] -= coordinate;
+
+	float64_t2 roots = hatchutils::getCurveRoot(points[0], points[1], points[2]);
+	if (roots.x >= 0.0 && roots.x <= 1.0) return roots.x;
+	if (roots.y >= 0.0 && roots.y <= 1.0) return roots.y;
+	return core::nan<double>();
+}
+
+float64_t2 Hatch::QuadraticBezier::evaluateBezier(double t) const
+{
+	float64_t2 position = p[0] * (1.0 - t) * (1.0 - t)
+		+ 2.0 * p[1] * (1.0 - t) * t
+		+ p[2] * t * t;
+	return position;
+}
+
+// https://pomax.github.io/bezierinfo/#pointvectors
+float64_t2 Hatch::QuadraticBezier::tangent(double t) const
+{
+	// TODO: figure out a tangent algorithm for when this becomes A, B, C
+	auto derivativeOrder1First = 2.0 * (p[1] - p[0]);
+	auto derivativeOrder1Second = 2.0 * (p[2] - p[1]);
+	auto tangent = (1.0 - t) * derivativeOrder1First + t * derivativeOrder1Second;
+	auto len = sqrt(tangent.x * tangent.x + tangent.y * tangent.y);
+	return tangent / len;
+}
+
+Hatch::QuadraticBezier Hatch::QuadraticBezier::splitCurveTakeLower(double t) const
```

Comment:
I won't also review this function and the next one, cause the math is already done in that web page and your tests show promise that these work.